spec/firebase_app_distribution_action_spec.rb

```diff
@@ -1,141 +1,147 @@
 describe Fastlane::Actions::FirebaseAppDistributionAction do
   let(:fake_file) { StringIO.new }
-
-  before do
+  let(:temp_connection) { double("Connection") }
+  let(:fake_get_app_response_valid) { double("Response", status: 200, body: { projectNumber: "1", appId: "1", platform: "android", bundleId: "1", contactEmail: "Hello@world.com" }) }
+  let(:fake_get_app_response_no_email) { double("Response", status: 200, body: { projectNumber: "1", appId: "1", platform: "android", bundleId: "1", contactEmail: "" }) }
+  let(:fake_get_app_response_invalid_app_id) { double("Response", status: 400, body: { error: { code: 400, message: "Request contains an invalid argument.", status: "INVALID_ARGUMENT" } }) }
+  let(:fake_get_app_response_permission_denied) { double("Response", status: 403, body: { error: { code: 403, message: "The caller does not have permission", status: "PERMISSION_DENIED" } }) }
+  let(:fake_binary) { double("Binary") }
+
+  before(:each) do
+    allow(Fastlane::Actions::FirebaseAppDistributionAction).to receive(:connection).and_return(temp_connection)
+    allow(fake_binary).to receive(:read).and_return("Hello World")
+
+    allow(File).to receive(:open).and_return(fake_binary)
     allow(Tempfile).to receive(:new).and_return(fake_file)
     allow(fake_file).to receive(:unlink)
```

Comment:
I notice a lot of mocking in this spec setup which can be a sign of a [change detector test](https://sites.google.com/a/google.com/tater-tott/posts/2014/09/episode_350__change_detector_tests_considered_harmful)

I know it might be challenging to avoid mocks in this case because of all the filesystem interaction, but you might be able to remove one layer of mocking by creating your own version of a MockFile which will behave a little bit more like a real object than relying on you to aggressively mock every interaction out.  As long as the fake mostly follows the API you're pretending to be you might find it more reusable across specs and require fewer `allow(x).to receive` calls to get it ready for use.  You can flesh out your mock class as needed when you find more unmocked interactions, and you might find you can use it for various types of files in your spec.

EG:

```
class MockFile
   def initialize(path, contents)
     @path = path
     @contents = contents || ""
   end

   def read
      @contents 
   end

   def write(text)
     @contents << text # or overwrite, not sure what the behavior here should be
   end

   def path
     @path
   end
end
```

Also other projects commonly use something like [fakefs](https://github.com/fakefs/fakefs) to do this, but given this is an open source project the drawbacks of pulling in another dependency might outweigh the benefits of getting some of this support in your specs.  Probably better to remain lightweight.

Reply:
A lot of the tests here are older tests that we will be deleting. I know you already looked at the unit test PR, but that has the format we will most likely use going forward.

Reply:
👍